xapian-applications/omega/handler_libebook.cc

```diff
@@ -0,0 +1,181 @@
+#include <libe-book/libe-book.h>
+
+#define PARSE_FIELD(START, END, FIELD, OUT) \
+   parse_metadata_field((START), (END), (FIELD), (CONST_STRLEN(FIELD)), (OUT))
+
+using libebook::EBOOKDocument;
+using namespace std;
+using namespace librevenge;
+
+static void
+parse_metadata_field(const char* start,
+		     const char* end,
+		     const char* field,
+		     size_t len,
+		     string& out)
+{
+    if (size_t(end - start) > len && memcmp(start, field, len) == 0) {
+	start += len;
+	while (start != end && isspace(*start)) start++;
+	if (start != end && (end[-1] != '\r' || --end != start)) {
+	    if (!out.empty())
+		out.push_back(' ');
+	    out.append(start, end - start);
+	}
+    }
+}
+
+static void
+parse_metadata(const char* data,
+	       size_t len,
+	       string& author,
+	       string& title,
+	       string& keywords)
+{
+    const char* p = data;
+    const char* end = p + len;
+
+    while (p != end) {
+	const char* start = p;
+	p = static_cast<const char*>(memchr(p, '\n', end - start));
+	const char* eol;
+	if (p)
+	    eol = p++;
+	else
+	    p = eol = end;
+	if ((end - start) > 5 && memcmp(start, "meta:", 5) == 0) {
+	    start += 5;
+	    switch (*start) {
+		case 'i': {
+		    if (author.empty())
+			PARSE_FIELD(start, eol, "initial-creator", author);
+		    break;
+		}
+		case 'k': {
+		    PARSE_FIELD(start, eol, "keyword", keywords);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 3 && memcmp(start, "dc:", 3) == 0) {
+	    start += 3;
+	    switch (*start) {
+		case 'c': {
+		    if (!author.empty())
+			author.clear();
+		    PARSE_FIELD(start, eol, "creator", author);
+		    break;
+		}
+		case 's': {
+		    PARSE_FIELD(start, eol, "subject", keywords);
+		    break;
+		}
+		case 't': {
+		    PARSE_FIELD(start, eol, "title", title);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 8 && memcmp(start, "dcterms:", 8) == 0) {
+	    start += 8;
+	    PARSE_FIELD(start, eol, "available", keywords);
+	}
+    }
+}
+
+static void
+clear_text(string& str, const char* text)
+{
+    for (int i = 0; text[i] != '\0'; ++i)
+	if (!isspace(text[i]) || (i && !isspace(text[i - 1])))
+	    str.push_back(text[i]);
+}
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages)
+{
+    try {
+	shared_ptr<RVNGInputStream> input;
+	RVNGString content_dump, metadata_dump;
+	const char* file = filename.c_str();
+	bool f_txt = false, f_meta = false;
+
+	if (RVNGDirectoryStream::isDirectory(file))
+	    input.reset(new RVNGDirectoryStream(file));
+	else
+	    input.reset(new RVNGFileStream(file));
+
+	EBOOKDocument::Type type = EBOOKDocument::TYPE_UNKNOWN;
+	EBOOKDocument::Confidence confidence =
+				EBOOKDocument::isSupported(input.get(), &type);
+
+	if (EBOOKDocument::CONFIDENCE_SUPPORTED_PART == confidence) {
+	    input.reset(RVNGDirectoryStream::createForParent(file));
+	    confidence = EBOOKDocument::isSupported(input.get(), &type);
+	}
```

Comment:
Are you sure we want this?

If I follow what this does, it means that if we try to index a file `/home/olly/Documents/foo.fb2` libe-book could tell us `CONFIDENCE_SUPPORTED_PART` and then we'd ask libe-book to try to extract the directory `/home/olly/Documents` instead.  Given how omindex works by recursing a directory structure, this doesn't seem like it's ever going to be a good idea.

Reply:
Sorry, misunderstand the 'createForParent' function. I am going to remove it.

Reply:
I think this mechanism is probably there for the case of things like Apple keynote where the document can be a directory with a `.key` extension.  If you ask libetonyek to open a file within that such as `/path/to/foo.key/preview.jpg` it might return `CONFIDENCE_SUPPORTED_PART` and `createForParent` would then open `/path/to/foo.key`.